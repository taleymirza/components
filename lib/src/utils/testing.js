import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(mount(component)).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    const dest = computed[key] || (computed[key] = {})
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return div.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}
