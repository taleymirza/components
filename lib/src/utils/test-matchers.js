import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toSetDefaultTheme(Component) {
    const wrapper = mount(<Component />)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  }
})
