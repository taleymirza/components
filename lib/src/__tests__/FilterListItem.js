/* eslint-disable jsx-a11y/anchor-has-content, jsx-a11y/anchor-is-valid */
import React from 'react'
import FilterList from '../FilterList'
import {render} from '../utils/testing'
import {COMMON} from '../constants'

describe('FilterList.Item', () => {
  it('has default theme', () => {
    expect(FilterList.Item).toSetDefaultTheme()
  })

  it('implements system props', () => {
    expect(FilterList.Item).toImplementSystemProps(COMMON)
  })

  it('renders an <a> by default', () => {
    expect(render(<FilterList.Item />).type).toEqual('a')
  })

  it('renders the given "as" prop', () => {
    const Type = props => <b {...props} />
    expect(render(<FilterList.Item as={Type} />)).toMatchSnapshot()
  })

  it('respects the "selected" prop', () => {
    expect(render(<FilterList.Item selected />)).toMatchSnapshot()
  })

  it('respects "count" prop', () => {
    const CountMock = render(<FilterList.Item count="400" />).children.pop()
    expect(CountMock.type).toEqual('span')
  })
})
