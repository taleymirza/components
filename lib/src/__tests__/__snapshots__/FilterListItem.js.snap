// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterList.Item renders the given "as" prop 1`] = `
.c0 {
  position: relative;
  display: block;
  padding: 8px 11px;
  margin: 0 0 5px 0;
  overflow: hidden;
  font-size: 14px;
  color: #586069;
  background-color: !important;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  border-radius: 3px;
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #eaecef;
}

.c0:active {
  color: #fff;
  background-color: #0366d6;
}

.c0 .count {
  float: right;
  font-weight: 600;
}

<b
  className="c0"
  theme={
    Object {
      "borders": Array [
        0,
        "1px solid",
      ],
      "breakpoints": Array [
        "544px",
        "768px",
        "1012px",
        "1280px",
      ],
      "colors": Object {
        "black": "#1b1f23",
        "blackfade15": "rgba(27, 31, 35, 0.15)",
        "blackfade20": "rgba(27, 31, 35, 0.20)",
        "blackfade35": "rgba(27, 31, 35, 0.35)",
        "blackfade50": "rgba(27, 31, 35, 0.5)",
        "blue": Array [
          "#f1f8ff",
          "#dbedff",
          "#c8e1ff",
          "#79b8ff",
          "#2188ff",
          "#0366d6",
          "#005cc5",
          "#044289",
          "#032f62",
          "#05264c",
        ],
        "bodytext": "#24292e",
        "button": Object {
          "activeBg": "rgb(233, 236, 239)",
          "bg2": "rgb(239, 243, 246)",
          "border": "rgba(27, 31, 35, 0.2)",
          "dangerActiveBg": "rgb(181, 32, 44)",
          "dangerFocusShadow": "rgba(203, 36, 49, 0.4)",
          "dangerHoverBgImage": "rgb(222, 68, 80)",
          "disabledColor": "rgba(36, 41, 46, 0.4)",
          "focusShadow": "rgba(3, 102, 214, 0.3)",
          "hoverBg": "rgb(230, 235, 241)",
          "outlineBlue": "rgb(3, 102, 214)",
          "outlineShadow": "rgba(3, 102, 214, 0.4)",
          "primaryActiveBg": "rgb(39, 159, 67)",
          "primaryActiveShadow": "rgba(27, 31, 35, 0.15)",
          "primaryBg": "rgb(40, 167, 69)",
          "primaryBgImage": "rgb(52, 208, 88)",
          "primaryBorder": "rgba(27, 31, 35, 0.5)",
          "primaryFocusShadow": "rgba(52, 208, 88, 0.4)",
          "primaryHoverBg": "rgb(38, 159, 66)",
          "primaryHoverBgImage": "rgb(47, 203, 83)",
          "white": "rgb(255, 255, 255)",
        },
        "counter": Object {
          "bg": "rgba(27, 31, 35, 0.08)",
        },
        "filterList": Object {
          "hoverBg": "#eaecef",
        },
        "gray": Array [
          "#fafbfc",
          "#f6f8fa",
          "#e1e4e8",
          "#d1d5da",
          "#959da5",
          "#6a737d",
          "#586069",
          "#444d56",
          "#2f363d",
          "#24292e",
        ],
        "green": Array [
          "#f0fff4",
          "#dcffe4",
          "#bef5cb",
          "#85e89d",
          "#34d058",
          "#28a745",
          "#22863a",
          "#176f2c",
          "#165c26",
          "#144620",
        ],
        "orange": Array [
          "#fff8f2",
          "#ffebda",
          "#ffd1ac",
          "#ffab70",
          "#fb8532",
          "#f66a0a",
          "#e36209",
          "#d15704",
          "#c24e00",
          "#a04100",
        ],
        "purple": Array [
          "#f5f0ff",
          "#e6dcfd",
          "#d1bcf9",
          "#b392f0",
          "#8a63d2",
          "#6f42c1",
          "#5a32a3",
          "#4c2889",
          "#3a1d6e",
          "#29134e",
        ],
        "red": Array [
          "#ffeef0",
          "#ffdce0",
          "#fdaeb7",
          "#f97583",
          "#ea4a5a",
          "#d73a49",
          "#cb2431",
          "#b31d28",
          "#9e1c23",
          "#86181d",
        ],
        "state": Object {
          "error": "#d73a49",
          "failure": "#d73a49",
          "pending": "#dbab09",
          "queued": "#dbab09",
          "success": "#28a745",
          "unknown": "#959da5",
        },
        "white": "#fff",
        "whitefade15": "rgba(255, 255, 255, 0.15)",
        "yellow": Array [
          "#fffdef",
          "#fffbdd",
          "#fff5b1",
          "#ffea7f",
          "#ffdf5d",
          "#ffd33d",
          "#f9c513",
          "#dbab09",
          "#b08800",
          "#735c0f",
        ],
      },
      "fontSizes": Array [
        12,
        14,
        16,
        20,
        24,
        32,
        40,
        48,
      ],
      "fontWeights": Object {
        "bold": 600,
        "light": 300,
        "normal": 400,
      },
      "fonts": Object {
        "mono": "SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, Courier, monospace",
        "normal": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      },
      "lineHeights": Object {
        "condensed": 1.25,
        "condensedUltra": 1,
        "default": 1.5,
      },
      "maxWidths": Object {
        "large": "1012px",
        "medium": "768px",
        "small": "544px",
        "xlarge": "1280px",
      },
      "radii": Array [
        0,
        3,
        6,
      ],
      "shadows": Object {
        "extra-large": "0 10px 50px rgba(27, 31, 35, 0.07)",
        "formControl": "rgba(27, 31, 35, 0.075) 0px 1px 2px inset",
        "formControlFocus": "rgba(3, 102, 214, 0.3) 0px 0px 0px 0.2em",
        "large": "0 1px 15px rgba(27, 31, 35, 0.15)",
        "medium": "0 1px 5px rgba(27, 31, 35, 0.15)",
        "small": "0 1px 1px rgba(27, 31, 35, 0.1)",
      },
      "space": Array [
        0,
        4,
        8,
        16,
        24,
        32,
        40,
        48,
        64,
        80,
        96,
        112,
        128,
      ],
    }
  }
/>
`;

exports[`FilterList.Item respects the "selected" prop 1`] = `
.c0 {
  position: relative;
  display: block;
  padding: 8px 11px;
  margin: 0 0 5px 0;
  overflow: hidden;
  font-size: 14px;
  color: #fff;
  background-color: #0366d6!important;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  border-radius: 3px;
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #eaecef;
}

.c0:active {
  color: #fff;
  background-color: #0366d6;
}

.c0 .count {
  float: right;
  font-weight: 600;
}

<a
  className="c0"
  selected={true}
/>
`;
